staging/src/k8s.io/apiserver/pkg/server/options/encryptionconfig/config.go

```diff
@@ -513,3 +515,130 @@ func (u unionTransformers) TransformFromStorage(ctx context.Context, data []byte
 func (u unionTransformers) TransformToStorage(ctx context.Context, data []byte, dataCtx value.Context) (out []byte, err error) {
 	return u[0].TransformToStorage(ctx, data, dataCtx)
 }
+
+// DynamicTransformers holds transformers that are dynamically updated.
+type DynamicTransformers struct {
+	transformTracker *atomic.Value
+	SwapNotification chan bool
+}
+
+type transformTracker struct {
+	transformerOverrides map[schema.GroupResource]value.Transformer
+	closeTransformers    context.CancelFunc
+	inFlight             *atomic.Int64
+	closing              *atomic.Bool
+}
+
+// NewDynamicTransformers initializes DynamicTransformers when api-server starts.
+func NewDynamicTransformers(
+	transformerOverrides map[schema.GroupResource]value.Transformer,
+	closeTransformers context.CancelFunc,
+) *DynamicTransformers {
+	dynamicTransformers := &DynamicTransformers{
+		transformTracker: &atomic.Value{},
+	}
+
+	tracker := &transformTracker{
+		transformerOverrides: transformerOverrides,
+		closeTransformers:    closeTransformers,
+		inFlight:             &atomic.Int64{},
+		closing:              &atomic.Bool{},
+	}
+	dynamicTransformers.transformTracker.Store(tracker)
+
+	dynamicTransformers.SwapNotification = make(chan bool)
+
+	return dynamicTransformers
+}
+
+// TransformerForResource returns the transformer for the given resource.
+func (d *DynamicTransformers) TransformerForResource(resource schema.GroupResource) value.Transformer {
+	return &resourceTransformer{
+		resource:         resource,
+		transformTracker: d.transformTracker,
+	}
+}
+
+// Set sets the transformer overrides. This method is not go routine safe and must only be called by the same, single caller throughout the lifetime of this object.
+func (d *DynamicTransformers) Set(
+	transformerOverrides map[schema.GroupResource]value.Transformer,
+	closeTransformers context.CancelFunc,
+) {
+	// store new values
+	newTransformTracker := &transformTracker{
+		transformerOverrides: transformerOverrides,
+		closeTransformers:    closeTransformers,
+		inFlight:             &atomic.Int64{},
+		closing:              &atomic.Bool{},
+	}
+
+	oldTransformTracker := d.transformTracker.Swap(newTransformTracker).(*transformTracker)
+
+	oldTransformTracker.closing.Store(true)
+	if oldTransformTracker.inFlight.Load() == 0 {
+		oldTransformTracker.closeTransformers()
+	}
+
+	// send swap complete signal
+	d.SwapNotification <- true
```

Comment:
@enj Is there a better way to do this? I tried returning a function with `<-chan struct{}`, but that was getting convoluted as we need to do this for every `.Set` call.

Reply:
Removed and implemented with another approach